Fixes #12235: handle dangling reference during VHDS update request

dmitri-d opened the pull request:
Fixes https://github.com/envoyproxy/envoy/issues/12235

Signed-off-by: Dmitri Dolguikh <ddolguik@redhat.com>

Risk Level: low
Testing: existing tests
Docs Changes: none
Release Notes: none

dmitri-d commented:
ping @stevenzzzz

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
+  // We use weak pointers here as RdsRouteConfigProviderImpl instance could go away before the
+  // dispatcher had a chance to execute the callback.
+  factory_context_.dispatcher().post(
+      [sub = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
```
I think you only need the config_cbs to help decide if the config provider is gone.

dmitri-d replied:
Yes, but it's dependent on the position of the member var in the source file. I figured this is the safest?

stevenzzzz replied:
I dont get it: as you pointed out in the other comment, the RDS provider is guaranteed to be allocate/deallocated on main thread. so if we can lock config_cbs here, that means the subscription is definitely alive. this lambda is run in main thread, where deallocation of RDS provider happens in sequence with this lambda.

dmitri-d replied:
You are right, fixed

stevenzzzz replied:
sorry, I think you want to check the config_cbs_ actually. 
Subscription is shared between routeconfigProviders, there could be a case where this route provider could be gone, but the subscription stays.

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
+       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
+       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
+        auto s = sub.lock();
+        auto cb = config_cbs.lock();
```
use a better name pls?

dmitri-d replied:
renamed

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
+        auto cb = config_cbs.lock();
+        if (s && cb) {
+          s->updateOnDemand(alias);
+          cb->push_back({alias, thread_local_dispatcher, route_config_updated_cb});
```
if in main thread, management server already sent back response with the alias, do you want to send another on demand request to management server?

Depends on management server implementation, if it only sends a response when there is some update of the project, then this HttpRequest will hang till timeout, while Envoy has the config to proceed.

dmitri-d replied:
This gets pretty complicated quickly: the only time (that I can see) an instance of `RdsRouteConfigProviderImpl`can be deallocated is when LDS is used and the listener configuration has changed. At this point all connections are going to be drained, trying to update a vhost seems not worth it to me.

Even if we wanted to do that, we'd still need an active subscription (may not be there yet, and no way to find out which one to use) and a place to store the callback to notify the on_demand_update filter (which may not be there any more, as the active stream it belongs to may have been closed).

stevenzzzz replied:
I am describing a scenario where no rds subscription destruction is involved. 

could the management server send required vhost to Envoy before the on-demand request is sent? in that case, the management server may or may not send the requested resource again depends on its implementation, in which case the callback will not be called and the request will only timeout.  this is very implementation specific, so I am not too concerned.

dmitri-d commented:
* added a test

dmitri-d commented:
* merged in changes from master

stevenzzzz commented on test/common/router/rds_impl_test.cc:
```diff
@@ -278,6 +279,39 @@ TEST_F(RdsImplTest, FailureSubscription) {
+    EXPECT_CALL(server_factory_context_.dispatcher_, post(_)).WillOnce(SaveArg<0>(&post_cb));
+    rds->requestVirtualHostsUpdate("testing", local_thread_dispatcher,
+                                   std::make_shared<Http::RouteConfigUpdatedCallback>(
+                                       Http::RouteConfigUpdatedCallback([](bool) -> void {})));
```
nit: might worth check if the lambda is actually not called. 

The dangling reference may not lead to an Exception but some segv error.

dmitri-d commented:
ping @stevenzzzz.

stevenzzzz commented:
sorry I thought I committed the comment.

there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_.

dmitri-d commented:
> there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_

In this context, a weak_ptr to either the subscription or the list of callbacks would work; if anything the subscription is longer-lived and has a benefit of being already stored in a shared_ptr.

dmitri-d commented:
* switched to use config_update_callbacks to determine if the instance of RdsRouteConfigProviderImpl is still alive.

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
+      [subscription = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
+       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
+       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
+        auto sub = subscription.lock();
```
you can move this into the if statement below.

stevenzzzz replied:
could you also move this sub lock to the if statement below?

stevenzzzz commented:
> > there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_
> 
> In this context, a weak_ptr to either the subscription or the list of callbacks would work; if anything the subscription is longer-lived and has a benefit of being already stored in a shared_ptr.
 if config_callbacks are gone, it's undefined behavior again when you push a callback into the list.

Thanks for the fix.
LGTM

dmitri-d commented:
* responded to feedback

dmitri-d commented:
* responded to feedback

dmitri-d commented:
Looks like 1rm build failure is due to a flaky test.

dmitri-d commented:
ping @stevenzzzz

stevenzzzz commented:
I LGTMed already.
Let's ping @snowp   
😄

dmitri-d commented:
* responded to feedback
* merged in master

ping @snowp

stevenzzzz commented:
/assign htuch@ 
FYI Harvey

stevenzzzz commented:
/assign @htuch 
FYI

mattklein123 commented on source/common/router/rds_impl.cc:
```diff
@@ -306,11 +306,17 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // The RdsRouteConfigProviderImpl instance can go away before the dispatcher has a chance to
+  // execute the callback. We use a weak_ptr here to verify that the instance is still around.
```
Why don't we allocate the provider itself with shared_from_this and then we can capture a weak reference to ourselves and post that? This seems a bit strange and fragile?

stevenzzzz replied:
I'd hesitate to do so, with that we could get into another rabbit hole where a worker thread may accidentally get a shared_ptr of the provider and introduce another bunch of life cycle problem due to provider can only be con/destructed on main thread. 
thoughts?

mattklein123 replied:
OK fair enough. Can we add a comment on why we are not doing that so the next person that comes along doesn't ask the same question? Thank you.

/wait

dmitri-d replied:
Ahh, yes, this is better.

dmitri-d commented:
* switched to using a shared_from_this instance of a RdsRouteConfigProviderImpl
 * merged in latest master

stevenzzzz commented on source/common/router/rds_impl.h:
```diff
@@ -193,7 +193,8 @@ struct UpdateOnDemandCallback {
 class RdsRouteConfigProviderImpl : public RouteConfigProvider,
-                                   Logger::Loggable<Logger::Id::router> {
+                                   Logger::Loggable<Logger::Id::router>,
+                                   public std::enable_shared_from_this<RdsRouteConfigProviderImpl> {
```
can we please not do this. When provider is capable of shared_from_this, we actual open its access to worker threads, which I'd hope we can avoid: providers are supposed to be created/owned by HCM and no one else should have ownership to it, otherwise it'd introduce lots of trouble to do destruction orders when Envoy tears down. 
If you need a cleaner signal to decide if we need a post, one thing you can do is to lock on route_config_updated_cb, which the finest grained object that tells if the source httprequest is alive or not.

mattklein123 replied:
I don't quite understand this now that I think about it more. This is already a shared_ptr allocated object. Why does it matter if it enabled shared_from_this?

stevenzzzz replied:
I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

dmitri-d replied:
I think the change is quite safe: we aren't changing provider interface, just the concrete implementation: to create another shared_ptr one would need to cast a RouteConfigProvider object to RdsRouteConfigProviderImpl before being able to do so. Code review would catch such a cast or a call to `shared_from_this()`. I can add a note re: intended usage of shared_from_this if that would help.

mattklein123 replied:
I agree that this is the right change modulo any other lifetime issue.

stevenzzzz replied:
it's a tricky case and it's not about life cycle of provider when the post lambda gets called.
 the problem is that The ownership of any "tls instance owner" should not be spilled into other threads than the main thread.
please see my other reply why I am against this, your change is introducing another race which could ends in an assertion failure.

stevenzzzz commented on source/common/router/rds_impl.cc:
```diff
@@ -308,15 +308,16 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
-      });
+  factory_context_.dispatcher().post([this,
+                                      self = std::weak_ptr<RdsRouteConfigProviderImpl>(
+                                          shared_from_this()),
```
this could cause trouble as it introduces another race between worker threads and main thread: for example on envoy teardown, the lambda created in the worker thread may be the last one to be teardown, in which case, the destruction of the provider will happen in worker thread, which triggers an assert failure because it holds an TLS instance which can only be destructed in main thread.

mattklein123 replied:
But how is this safe then? We capture "this" so it can still crash?

stevenzzzz replied:
I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

mattklein123 replied:
> i am thinking capturing a member of the provider will guarantee that the provider is alive when this cb is called inside main thread.

It definitely does not guarantee that. It just prevents the member from going away.

stevenzzzz replied:
let me rephrase why this change is a regression: 
  1. this lambda "L" is created with a shared pointer to the RDS config provider, and it's in worker thread.
  2. it's destructed in worker thread after it's posted to main thread, 
  3. in main thread, the rds config provider could go away because of LDS or Envoy teardown, in which case the ownership of the config provider will fall to the "L" if it's not deleted from the worker thread yet.
  4. deallocation of the RDS config provider in worker thread will cause an assertion failure, which is our fence on clear boundary between config pipeline and data plane, as the tls instance held by the provider can only be destructed in main thread.

mattklein123 replied:
Yes I agree ^ is a problem, but this code is still broken. If the provider goes away on the main thread, we still capture this. That is not correct. I think given how long we are talking about this we probably need a cleaner solution here. If the original change is actually safe can the comments be clearly updated to why?

stevenzzzz replied:
Right. So if we capture a weak_ptr of the callback list, and verify it's alive in main thread, we can make sure that we are accessing an active provider(and its subscription), meanwhile, the provider ownership is still on main thread and we dont need to worry about the race mentioned above.

trying to lock on the route_config_updated_cb will also do the trick, but it has a different semantic: if the callback is alive, which means the hcm is alive and the provider associated with it will be alive. 
but now we are firing an on-demand when the requesting HttpRequest is alive, if we lock on provider, we are firing as long as there was a request(may be gone).

mattklein123 replied:
Yes, OK, I think this make sense. @dmitri-d can we revert back to what you had before but also add a lot more comments? I would like to get this merged as I'm dealing with additional issues in this area and I want to fix on top of this depending on if anything else needs fixing. Thank you!

/wait

htuch replied:
Do we need to capture a weak_ptr to both subscription and callback list? Or would just subscription work? I think the original solution looks right, but it seems awkward to have to lock two objects. Better documentation on lifetime guarantees and relationships would definitely help. I think what you really want is some sentinel member element that provides "am I alive?"; normally this could be `shared_from_this` but this doesn't work for the reasons @stevenzzzz points out.

mattklein123 commented:
I just checked and the previous version of this patch doe *not* fix https://github.com/envoyproxy/envoy/issues/12235#issuecomment-671634502, so I will continue to debug.

mattklein123 commented:
Nevermind regarding ^ this is the same TSAN issue that is fixed with instrumented TSAN which I am not using. :(

dmitri-d commented:
*  Reverted shared_from_this changes
* Introduced a dedicated 'still alive' flag: functionally the same as relying on a weak_ptr to config_update_callbacks_, but reads a bit clearer (I think). If everyone is ok with this, I'll add a comment explaining the lifecycle there.

stevenzzzz commented:
> This makes sense to me, @stevenzzzz @mattklein123 ?

LGTM, making a shared_ptr on a callback list indeed looks weird.

dmitri-d commented:
* fixed formatting
* added a comment explaining the need for `still_alive` flag
